### PR TITLE
Component-wise MCMC for subset simulation

### LINKED ISSUE #165: We're not using Modified (component-wise) MCMC in subset simulation
In the docstrings of `SubSetSimulation`, we state that we're using the algorithm described in:

S.-K. Au and J. L. Beck. Estimation of Small Failure Probabilities in High Dimensions by Subset Simulation. [Probabilistic Engineering Mechanics 16, 263–277 (2001)](https://doi.org/10.1016/S0266-8920(01)00019-4)

A modified MCMC algorithm, which samples component wise, allowing for high dimensional problems.

Instead we've implemented standard MCMC, which struggles in high dimensions due to small densities of the proposal distribution, leading to small acceptance rates.

We should implement the modified version also, and change these docstrings.

### PR DESCRIPTION
**Pull request by AnderGray**
This PR changes the MCMC sampler in `SubSetSimulation` from usual Metropolis-Hastings, to component-wise MH. Dimensions are accepted independently, giving a better acceptance rate in higher dimensions.

closes #165

### CODE REVIEW

File: src/simulations/subset.jl [hunk 1]
```diff
@@ -94,22 +94,29 @@ end
...
-of levels and `λ` (λ = 1 recommended) is the initial scaling parameter and `Na` is the number of
-times to update `λ` per subset level (number of partitions of seeds). The initial variance of the proposal distribution is `λ`.
+of levels, `λ` (λ = 1 recommended) is the initial scaling parameter, and `Na` is the number simulations that will be run before `λ` 
+is updated. Note that Na must be a multiple of n * target: `mod(ceil(n * target), Na) == 0)`. The initial variance of the proposal distribution is `s`.
 
 
 Idea behind this algorithm is to adaptively select the correlation parameter of `s`
-at each intermediate level, by simulating a subset N_a of the chains
+at each intermediate level, by simulating a subset Na of the chains
 (which must be choosen without replacement at random) and modifying the acceptance rate towards the optiming
```
Comment (FriesischScott) [inline src/simulations/subset.jl:105]: might as well fix the tyypo *optiming*

**Review by FriesischScott — CHANGES_REQUESTED** [review]
Not sure what the issue with the actions is. Apart from some very minor suggestions this looks  good to me.

**Comment by AnderGray** [comment]
Thanks for taking a look, everything should be addressed

**Comment by AnderGray** [comment]
Potential bug found. Running this several times

```julia
using UncertaintyQuantification

Num_dim = 2
N_samples = 20

for j = 1:Num_dim                                                           #hide
    eval(:($(Symbol(:X,j)) = RandomVariable(Normal(0, 1), (Symbol(:X,$j))) ))#hide
end#hide

inputs = [eval(:($(Symbol(:X,i)))) for i = 1:Num_dim]                       #hide

y = Model(                                                                  #hide
    df -> sum.(eachrow(df[:, names(inputs)])),#hide
    :y,#hide
)#hide

pf_target = 10^-3

fail_limit = quantile(Normal(0, sqrt(Num_dim)), 1 - pf_target)#hide

function limitstate(df)#hide
    return fail_limit .- reduce(vcat, df.y)#hide
end#hide

simulation_method_1 = SubSetSimulation(N_samples, 0.1, 20, Uniform(-0.5, 0.5))#hide

pf_1, std_1, _ = probability_of_failure(y, limitstate, inputs, simulation_method_1)#hide
```

gives: 

```
ERROR: ArgumentError: invalid index: false of type Bool
Stacktrace:
 [1] to_index(i::Bool)
   @ Base ./indices.jl:293
 [2] to_index(A::Vector{Int64}, i::Bool)
   @ Base ./indices.jl:277
 [3] _to_indices1(A::Vector{Int64}, inds::Tuple{}, I1::Bool)
   @ Base ./indices.jl:359
 [4] to_indices
   @ ./indices.jl:354 [inlined]
 [5] to_indices
   @ ./indices.jl:350 [inlined]
 [6] getindex(::Vector{Int64}, ::Bool)
   @ Base ./abstractarray.jl:1294
 [7] nextlevelsamples(samples::DataFrames.DataFrame, performance::Vector{…}, threshold::Float64, models::Model, performancefunction::typeof(limitstate), inputs::Vector{…}, sim::SubSetSimulation)
   @ UncertaintyQuantification ~/Documents/julia/UQ_Fork_SS/Fork_2/UncertaintyQuantification.jl/src/simulations/subset.jl:300
 [8] probability_of_failure(models::Model, performancefunction::typeof(limitstate), inputs::Vector{RandomVariable}, sim::SubSetSimulation)
   @ UncertaintyQuantification ~/Documents/julia/UQ_Fork_SS/Fork_2/UncertaintyQuantification.jl/src/simulations/subset.jl:215
 [9] top-level scope
   @ REPL[35]:1
Some type information was truncated. Use `show(err)` to see complete types.

```

**Comment by AnderGray** [comment]
I've also encountered issue #138

**Comment by FriesischScott** [comment]
> Potential bug found. Running this several times
> 
> ```julia
> using UncertaintyQuantification
> 
> Num_dim = 2
> N_samples = 20
> 
> for j = 1:Num_dim                                                           #hide
>     eval(:($(Symbol(:X,j)) = RandomVariable(Normal(0, 1), (Symbol(:X,$j))) ))#hide
> end#hide
> 
> inputs = [eval(:($(Symbol(:X,i)))) for i = 1:Num_dim]                       #hide
> 
> y = Model(                                                                  #hide
>     df -> sum.(eachrow(df[:, names(inputs)])),#hide
>     :y,#hide
> )#hide
> 
> pf_target = 10^-3
> 
> fail_limit = quantile(Normal(0, sqrt(Num_dim)), 1 - pf_target)#hide
> 
> function limitstate(df)#hide
>     return fail_limit .- reduce(vcat, df.y)#hide
> end#hide
> 
> simulation_method_1 = SubSetSimulation(N_samples, 0.1, 20, Uniform(-0.5, 0.5))#hide
> 
> pf_1, std_1, _ = probability_of_failure(y, limitstate, inputs, simulation_method_1)#hide
> ```
> 
> gives:
> 
> ```
> ERROR: ArgumentError: invalid index: false of type Bool
> Stacktrace:
>  [1] to_index(i::Bool)
>    @ Base ./indices.jl:293
>  [2] to_index(A::Vector{Int64}, i::Bool)
>    @ Base ./indices.jl:277
>  [3] _to_indices1(A::Vector{Int64}, inds::Tuple{}, I1::Bool)
>    @ Base ./indices.jl:359
>  [4] to_indices
>    @ ./indices.jl:354 [inlined]
>  [5] to_indices
>    @ ./indices.jl:350 [inlined]
>  [6] getindex(::Vector{Int64}, ::Bool)
>    @ Base ./abstractarray.jl:1294
>  [7] nextlevelsamples(samples::DataFrames.DataFrame, performance::Vector{…}, threshold::Float64, models::Model, performancefunction::typeof(limitstate), inputs::Vector{…}, sim::SubSetSimulation)
>    @ UncertaintyQuantification ~/Documents/julia/UQ_Fork_SS/Fork_2/UncertaintyQuantification.jl/src/simulations/subset.jl:300
>  [8] probability_of_failure(models::Model, performancefunction::typeof(limitstate), inputs::Vector{RandomVariable}, sim::SubSetSimulation)
>    @ UncertaintyQuantification ~/Documents/julia/UQ_Fork_SS/Fork_2/UncertaintyQuantification.jl/src/simulations/subset.jl:215
>  [9] top-level scope
>    @ REPL[35]:1
> Some type information was truncated. Use `show(err)` to see complete types.
> ```

Can you check what the state of `reject` is when this happens? Might be an edge case where no additional rejection has to be performed. 

Could also potentially be related to the very low number of samples  with only two seeds per level.

**Review by FriesischScott — APPROVED** [review]
Good work!

**Pull request merged**